test/distributed/org/apache/cassandra/distributed/test/sai/PartialUpdateHandlingTest.java

```diff
@@ -0,0 +1,401 @@
+            if (specification.restrictPartitionKey)
+                restricted.add("pk");
+
+            List<String> clauses = restricted.stream().map(c -> c + " = " + cells.get(c)).collect(Collectors.toList());
+            select.append(String.join(" AND ", clauses));
+
+            if (specification.readCL == QUORUM)
+            {
+                // Make sure node 2, which is the only complete replica, can't participate in a query...
+                CLUSTER.get(2).runOnInstance(() -> {
+                    Keyspace keyspace = Objects.requireNonNull(Schema.instance.getKeyspaceInstance(KEYSPACE));
+                    SecondaryIndexManager sim = keyspace.getColumnFamilyStore(TEST_TABLE_NAME).indexManager;
+                    for (String s : Arrays.asList("ck_idx", "s_idx", "a_idx", "b_idx"))
+                        sim.makeIndexNonQueryable(sim.getIndexByName(s), Index.Status.BUILD_FAILED);
+                });
+
+                SAIUtil.waitForIndexNonQueryable(CLUSTER, KEYSPACE, 2);
+            }
+
+            Object[][] result = CLUSTER.coordinator(1).execute(select.toString(), specification.readCL);
+
+            if (specification.readCL == QUORUM)
+            {
+                CLUSTER.get(2).runOnInstance(() -> {
+                    Keyspace keyspace = Objects.requireNonNull(Schema.instance.getKeyspaceInstance(KEYSPACE));
+                    SecondaryIndexManager sim = keyspace.getColumnFamilyStore(TEST_TABLE_NAME).indexManager;
+                    for (String s : Arrays.asList("ck_idx", "s_idx", "a_idx", "b_idx"))
+                        sim.makeIndexQueryable(sim.getIndexByName(s), Index.Status.BUILD_SUCCEEDED);
+                });
+
+                SAIUtil.waitForIndexQueryable(CLUSTER, KEYSPACE);
+            }
+
+            return result;
+        }
+    }
+
+    @Parameterized.Parameter
+    public Specification specification;
+
+    @Parameterized.Parameters(name = "{0}")
+    public static List<Object[]> parameters()
+    {
+        List<Object[]> parameters = new ArrayList<>();
+        
+        // Each test scenario operates over a different partition key.
+        int partitionKey = 0;
+
+        for (ConsistencyLevel readCL : new ConsistencyLevel[] { ALL, QUORUM })
+        {
+            for (boolean restrictPartitionKey : new boolean[] { false, true })
+            {
+                for (boolean flushPartials : new boolean[] { false, true })
+                {
+                    for (String[] columns : new String[][] { { "ck", "a" }, { "ck", "s" }, { "s", "a" }, { "a", "b" }, { "a" } })
+                        for (boolean existing : new boolean[] { false, true })
+                            parameters.add(new Object[]{ new Specification(restrictPartitionKey, columns, readCL, existing, StatementType.UPDATE, partitionKey++, flushPartials) });
+
+                    for (String[] columns : new String[][] { { "s", "a" }, { "a", "b" }, { "a" } })
```

Comment:
I'm not entirely sure how valuable the first two cases are here. It might be that the `{ "a" }` and `{ "s" }` scenarios are enough.